Support incrementing/decrementing the number below the cursor

d-e-s-o opened the pull request:
Vim supports incrementing & decrementing the number below the cursor (or after it) via Ctrl-a and Ctrl-x, respectively. Given fish's Vi mode support, it makes sense to provide similar functionality when working on the command line, to provide a more natural environment for Vim users. With this change we add the necessary functionality.

Closes: #8320 

I tested this via https://github.com/d-e-s-o/fish-shell/commit/23d42a9812eb341b83840a9b418b448c396bb7b1 (and had it running for years), but the unit tests seem to require `fish` modification to work.

## TODOs:
<!-- Just check off what what we know been done so far. We can help you with this stuff. -->
- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst <!-- Don't document changes for completions inside CHANGELOG.rst, there are lot of such edits -->

d-e-s-o committed:
Support incrementing/decrementing the number below the cursor

Vim supports incrementing & decrementing the number below the cursor (or
after it) via Ctrl-a and Ctrl-x, respectively. Given fish's Vi mode
support, it makes sense to provide similar functionality when working on
the command line, to provide a more natural environment for Vim users.
With this change we add the necessary functionality.

Closes: #8320

krobelus commented on share/functions/fish_vi_key_bindings.fish:
```diff
@@ -1,3 +1,77 @@
+alias vi_dec 'vi_inc_dec dec'
+alias vi_inc 'vi_inc_dec inc'
```
probably this should be called fish_vi_inc though I can see how that's ugly.

krobelus commented on share/functions/fish_vi_key_bindings.fish:
```diff
@@ -1,3 +1,77 @@
+alias vi_inc 'vi_inc_dec inc'
+
+# TODO: Currently we do not support hexadecimal and octal values.
+function vi_inc_dec --description 'increment or decrement a number'
```
this should be namespaced like fish_vi_inc_dec

Would be great to include your tests from 23d42a9812eb341b83840a9b418b448c396bb7b1.

I wonder if there is a strong reason why "commandline" silently
fails when run from `config.fish` or `fish -C` etc.

I'd expect that `fish -ic 'commandline hello; commandline'` prints "hello".
Maybe that was not considered in
a32248277f0 (Make commandline state thread safe, 2021-07-20);
looking at this diff, it seems like before that commit, that would have worked fine (I didn't test that though).
If that commit was really only about thread safety, then changing this to always use the current state if `is_interactive_session()` seems fine.

An earlier commit 782a7397364 (Remove confusing and misleading error
message when trying to repaint the prompt during startup, 2007-08-02)
added a special case which we can probably remove then -- the commandline will always be initialized,
but starts at the empty string.

I'd also expect that `fish -C 'commandline echo'` does prefill the commandline with "echo".

So maybe we can improve things here.

Alternatively, we could rewrite the test as a real interactive one (`tests/pexpects/`)

d-e-s-o replied:
That's a lot of expectations :-) I've added a commit adjust the commandline logic. Not sure if that's what you had in mind. Please take a look.

krobelus commented on share/functions/fish_vi_key_bindings.fish:
```diff
@@ -1,3 +1,77 @@
+function vi_inc_dec --description 'increment or decrement a number'
+  # The cursor is zero based, but all string functions assume 1 to be
+  # the lowest index. Adjust accordingly.
+  set cursor (math (commandline --cursor) + 1)
```
use `set -l ...` to avoid overwriting global or universal variables

krobelus commented on share/functions/fish_vi_key_bindings.fish:
```diff
@@ -1,3 +1,77 @@
+  # The cursor is zero based, but all string functions assume 1 to be
+  # the lowest index. Adjust accordingly.
+  set cursor (math (commandline --cursor) + 1)
+  set line (commandline --current-buffer)
```
doesn't look like this works for multiline command lines.
We can probably fix this by adding `| string collect` inside every
command substitution that can produce multiline output.
AFAICT that's only this line and some of the "string sub" calls.

krobelus commented on share/functions/fish_vi_key_bindings.fish:
```diff
@@ -1,3 +1,77 @@
+  set line (commandline --current-buffer)
+
+  set candidate (string sub --start $cursor -- $line)
+  if set just_found (string match --ignore-case --regex '^-?[0-9]+' -- $candidate)
```
we'll want to declare `set -l just_found` above this line, so both branches always use this local

d-e-s-o replied:
Seems fine to just add `--local` there, no need to move?

krobelus commented on share/functions/fish_vi_key_bindings.fish:
```diff
@@ -1,3 +1,77 @@
+
+    for i in (seq (math -- $cursor - 1) -1 1)
+      set candidate (string sub --start $i --end $end -- $line)
+      if set just_found (string match --ignore-case --regex '^-?[0-9]+$' -- $candidate)
```
why the `--ignore-case`?

krobelus commented on share/functions/fish_vi_key_bindings.fish:
```diff
@@ -271,6 +345,12 @@ function fish_vi_key_bindings --description 'vi-like key bindings for fish'
+    # increment or decrement numbers with ctrl+x ctrl+a
+    #
+    bind -s --preset -M default \ca vi_inc
+    bind -s --preset -M default \cx vi_dec
```
use ctrl-a and ctrl-x notation

d-e-s-o committed:
builtins/commandline: Remove initialized state special casing

Remove special casing around the initialization state when session is
interactive. This makes constructs such as
  $ fish -ic 'commandline hello; commandline'

work as expected in that 'hello' is printed. See [0] for a discussion.

[0]: https://github.com/fish-shell/fish-shell/pull/11570#discussion_r2144544053

d-e-s-o commented:
Seems MacOS `seq` requires positive increment or something like that, causing the new tests to fail. Any `fish` alternative that we could use?

d-e-s-o commented:
> > Seems MacOS seq requires positive increment or something like that, causing the new tests to fail. Any fish alternative that we could use?
> 
> Not sure. I guess either use `set i (math $i - 1)` or `seq 10 | tac`, or something in between

Thanks for the suggestion. I think that worked...but if I am reading [the other test failures](https://github.com/fish-shell/fish-shell/actions/runs/15685021241/job/44185900433?pr=11570#step:7:519) correctly, it appears to me as if perhaps cursor handling on MacOS behaves differently than on Linux?

Specifically,
```
commandline '123'; commandline --cursor 2; fish_vi_inc
commandline --current-buffer
# CHECK: 124                                 -> succeeds

commandline '123'; commandline --cursor 2; fish_vi_dec
commandline --current-buffer
# CHECK: 122                                 -> succeeds

commandline '123'; commandline --cursor 3; fish_vi_inc
commandline --current-buffer
# CHECK: 123                                 -> fails; actual value 124

commandline '123'; commandline --cursor 3; fish_vi_dec
commandline --current-buffer
# CHECK: 123                                 -> fails; actual value 122
```

Given that the only difference to the above is that we set the cursor behind the actual string, my guess would be that on MacOS that's not really what is happening. I don't have MacOS, so I am just making an educated guess based observation. I removed these two tests for now, as they don't seem to add much value, in my opinion, but perhaps that points to an issue in `fish` itself?
